### PR TITLE
Initial draft of prerendering-state.md

### PR DESCRIPTION
**Pull request by bokand**
Adds a document describing a proposal for `document.prerendering` and
updates some of the text in browsing-context.md to match current state
of thinking.

This is something along the lines of a "minimal" proposal in #2.

### CODE REVIEW

File: prerendering-state.md
Comment (ianvollick) [inline prerendering-state.md:63]: The term "preview" might be a more intuitive way of describing prerendering-while-visible.

Comment (bokand) [reply prerendering-state.md:63]: I agree, but I think it makes sense for this be a signal that we're in a prerendering-browsing-context. If that's the case, I think it'd make sense to match the names of the two things;  "prerendering-browsing-context" has the same issue. Would it make sense to rename them both? @jeremyroman @domenic for their thoughts.

Comment (ianvollick) [reply prerendering-state.md:63]: Ah, I was only suggesting we might use the term preview in the prose describing what the new prerender state did, not changing the name of anything web-facing.

Comment (bokand) [reply prerendering-state.md:63]: Ah - true, but that does beg the question whether naming the web-facing APIs 'prerender' evokes the right meaning if we need to clarify in prose

File: browsing-context.md
Comment (domenic) [inline browsing-context.md:181]: I can work on this TODO after this is merged.

In particular, I'm now feeling that we should reframe this section to say something like:

- Use all these individual APIs!
- But, maybe it'd be a good idea to expose the underlying primitive that the user agent is using, i.e. the loading mode? If so, here's our proposal. We're not yet aware of any use cases though, so consider this on-hold.

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:9]:
```suggestion
doc](https://docs.google.com/document/d/1Xzw0k8DgltI2ohapuDKmjRZLv7NVrRFGusW8IBtiCT0/edit?usp=sharing)._
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:18]:
```suggestion
* Avoid running JavaScript animations
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:22]:
```suggestion
A previous iteration of prerendering used [page-visibility](https://w3c.github.io/page-visibility/) to denote a
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:30]:
```suggestion
The most straightforward way to do this would be to bring back the previously specified `document.visibilityState ==
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:36]:
```suggestion
ordinary (i.e. non-prerendering) page that's hidden would likely want to "avoid running JavaScript animations" and
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:41]:
```suggestion
with `visible` either. In that case, we do want to "run JavaScript animations" and "fetch large resources" since it'll
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:49]:
```suggestion
```js
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:70]:
```suggestion
```webidl
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:84]:
```suggestion
### Relationship to `visibilityState`
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:95]:
```suggestion
#### `visibilityState` as a termination signal
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:66]:
```suggestion
## `document.prerendering`
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:28]:
```suggestion
## Why not `visibilityState=='prerender'`?
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:99]:
```suggestion
terminated. See also [w3c/PageVisibility#59](https://github.com/w3c/page-visibility/issues/59) for related discussion.
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:101]: maybe "being put into a portal (via portal predecessor adoption)"?

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:108]: Maybe note that you think that can be done as a followup, and in the meantime pages won't handle the case of being put into a portal perfectly, and that's OK?

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:110]: Let's delete this section and roll it into the TODO you inserted above, which I commented on.

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:124]: I wonder if you want an "open questions" h2-level section, which contains "visibilityState as a termination signal", "relationship to other states", and "Naming bikeshed"?

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:148]:
```suggestion
particular, if we want this to solve [w3c/PageVisibility#59](https://github.com/w3c/page-visibility/issues/59) we'd have the
```

File: prerendering-state.md
Comment (domenic) [inline prerendering-state.md:135]: Good summary of the issues!

**Review by domenic — APPROVED** [review]
Generally looks good! I added nits as inline suggestions; only a few substantive comments.

File: prerendering-state.md
Comment (jeremyroman) [inline prerendering-state.md:150]: It might be worth mentioning how it does (or doesn't, IIUC) interact with bfcache.

File: browsing-context.md
Comment (jeremyroman) [reply browsing-context.md:181]: Totally open to hiding loading mode from the JS API if we think the use cases that matter are covered by narrower APIs, and adding it later on if we actually need it (and are better informed about what it should look like).

File: prerendering-state.md
Comment (bokand) [reply prerendering-state.md]: Done

File: prerendering-state.md
Comment (bokand) [reply prerendering-state.md]: Done

File: prerendering-state.md
Comment (bokand) [reply prerendering-state.md:114]: Good idea, done.

File: prerendering-state.md
Comment (bokand) [reply prerendering-state.md]: Done

File: browsing-context.md
Comment (bokand) [reply browsing-context.md]: I've been thinking of `document.prerendering` as being based on whether or not the document is being rendered in a prerendering browsing context. Assuming that's true, it's also an underlying primitive.

The loading-mode seems related but maybe orthogonal? Thought experiment: would it ever make sense to have an existing loading mode (other than `default`), or perhaps a new one in the future, that was used in a regular browsing context? Or using the 'default' loading mode in a prerendering browsing context (would we use that for same-origin prerender/portal)? This doesn't sound crazy to me so I think maybe `document.prerendering` isn't necessarily just sugar for `loadingMode != 'default'`

File: prerendering-state.md
Comment (bokand) [reply prerendering-state.md:140]: I've been thinking of `document.prerendering` as "is in prerendering browsing context". Does BFCache have any expected interaction with the concept of a prerendering browsing context introduced in this repo? I expect not (and in that case I'll add that here) but I'm not sure if we're thinking of reusing some of the rendering/permissions restrictions.

File: prerendering-state.md
Comment (bokand) [reply prerendering-state.md:125]: Thanks :)

File: prerendering-state.md
Comment (jeremyroman) [reply prerendering-state.md:140]: I think likely not. Though it's similar in some regards to prerendering, script can't observe that much because it gets stopped for the duration where that would apply. But I think it's non-obvious enough to note.

File: browsing-context.md
Comment (mfalken) [inline browsing-context.md:206]: if (!document.prerendering) ?

File: prerendering-state.md
Comment (mfalken) [inline prerendering-state.md:95]: remove?

File: prerendering-state.md
Comment (mfalken) [inline prerendering-state.md:46]: Hm, this is rather unfortunate since `prerender` was only removed from the spec in 2019 and WebKit just removed it from trunk a few weeks ago in https://bugs.webkit.org/show_bug.cgi?id=215851. It's unfortunate that visibilityState is not a boolean, but we don't want to add new types because sites are treating it as one. But the argument above about prerendering not being mutually exclusive with visibility and hidden makes sense.

File: prerendering-state.md
Comment (mfalken) [reply prerendering-state.md:140]: +1 could be good to mention this. My understanding also is that bfcached pages do not run script so can't observe document.prerendering.

**Review by mfalken — APPROVED** [review]
lgtm for what it's worth, with some comments.

**Review by jeremyroman — APPROVED** [review]
lgtm whenever others are happy

File: prerendering-state.md
Comment (bokand) [reply prerendering-state.md:63]: This is still outstanding but unresolved. I filed #12 to keep this in mind but I think we can punt on this for now and leave it as "prerendering" since it's much clearer to outsiders what we're talking about. As other use cases shape up we can revisit it in the future.

**Comment by bokand** [comment]
Thanks everyone!

I don't have write access so could someone hit "merge" for me?

**Pull request merged**
